### PR TITLE
Request Limiter

### PR DESCRIPTION
**Pull request by mpalmi**
This commit introduces two new adaptive concurrency limiters in Vault, which should handle overloading of the server during periods of untenable request rate. The limiter adjusts the number of allowable in-flight requests based on latency measurements performed across the request duration. This approach allows us to reject entire requests prior to doing any work and prevents clients from exceeding server capacity.

The limiters intentionally target two separate vectors that have been proven to lead to server over-utilization.

- Back pressure from the storage backend, resulting in bufferbloat in the WAL system. (enterprise)
- Back pressure from CPU over-utilization via PKI issue requests (specifically for RSA keys), resulting in failed heartbeats.

Storage constraints can be accounted for by limiting logical requests according to their http.Method. We only limit requests with write-based methods, since these will result in storage Puts and exhibit the aforementioned bufferbloat.

CPU constraints are accounted for using the same underlying library and technique; however, they require special treatment. The maximum number of concurrent pki/issue requests found in testing (again, specifically for RSA keys) is far lower than the minimum tolerable write request rate. Without separate limiting, we would artificially impose limits on tolerable request rates for non-PKI requests. To specifically target PKI issue requests, we add a new PathsSpecial field, called limited, allowing backends to specify a list of paths which should get special-case request limiting.

For the sake of code cleanliness and future extensibility, we introduce the concept of a LimiterRegistry. The registry proposed in this PR has two entries, corresponding with the two vectors above. Each Limiter entry has its own corresponding maximum and minimum concurrency, allowing them to react to latency deviation independently and handle high volumes of requests to targeted bottlenecks (CPU and storage).

In both cases, utilization will be effectively throttled before Vault reaches any degraded state. The resulting 503 - Service Unavailable is a retryable HTTP response code, which can be handled to gracefully retry and eventually succeed. Clients should handle this by retrying with jitter and exponential backoff. This is done within Vault's API, using the go-retryablehttp library.

Limiter testing was performed via benchmarks of mixed workloads and across a deployment of agent pods with great success.

### CODE REVIEW

File: vault/core.go [hunk 1]
```diff
@@ -1284,6 +1298,11 @@ func NewCore(conf *CoreConfig) (*Core, error) {
 		return nil, err
 	}
 
+	c.limiterRegistry = conf.LimiterRegistry
```
Comment (raskchanky) [inline vault/core.go:1301]: Your ENT PR also has `c.limiterRegistry.Logger = conf.Logger.Named("limits")` after this line. Was this omitted here on purpose?

Comment (mpalmi) [reply vault/core.go:1301]: Yeah, this should be set earlier in `createCore`, so I dropped it here.

**Review by raskchanky — COMMENTED** [review]
Your ENT PR also has changes to `vault/request_forwarding.go` which don't appear to be in this PR. Is that on purpose?

**Comment by mpalmi** [comment]
> Your ENT PR also has changes to `vault/request_forwarding.go` which don't appear to be in this PR. Is that on purpose?

Yes to this as well! That was a change I made while trying to debug some things. I decided to drop it here. We may have to bring it back if we find that we need changes in the gRPC handler, but until then, leaving it out.

**Review by raskchanky — APPROVED** [review]

**Pull request merged**